### PR TITLE
fix(compression): Switch to MultiGzDecoder instead of GzDecoder

### LINKED ISSUE #7061: aws_s3 input doesn't handle files properly
Seems like s3 doesn't handle big files. I'm trying to send 17M archive with logs generated by AWS ALB:

```sh
# re-emit real logs to test bucket
aws s3 cp s3://mambu-prod-alb-logs-us-east-1/***/sand/AWSLogs/***/elasticloadbalancing/us-east-1/2021/03/15/***_elasticloadbalancing_us-east-1_app.awseb-AWSEB-1MVD8OW91UMOH.a32a5508b8fdaa6b_20210315T2115Z_10.219.6.78_1h057jj9.log.gz test.log.gz
aws s3 cp test.log.gz s3://mambu-staging-alb-logs-us-east-2/`date +%s | sha256sum | base64 | head -c 32`.log.gz

# ensure that logs are in valid format
gunzip test.log.gz
wc -l test.log
272228
```

### Vector Version
```
brew install vector
vector --version
vector 0.12.1 (v0.12.1 x86_64-apple-darwin 2021-03-12)
```

### Vector Configuration File

```yaml
sources:
  logs-us-east-1:
    type: aws_s3
    region: us-east-2
    compression: gzip
    sqs:
      delete_message: true
      queue_url: https://sqs.us-east-2.amazonaws.com/***/alb-logs-us-east-2-sqs-kourier
      visibility_timeout_secs: 30

sinks:
  console:
    type: console
    inputs: ["logs-*"]
    target: stdout
    encoding:
      codec: text
```

### Debug Output
RUST_BACKTRACE=full vector -vvv  -c vector.yaml
https://gist.github.com/afoninsky/f2929e608b8fadbc94792b6426b7dda1

### Expected Behavior
- SQS message is removed from the queue
- all (270k+) log entities are displayed in the console

### Actual Behavior
- SQS message is removed from the queue
- only 2 entities are displayed (see debug output)

### Additional Context
1. was able to repeat such behaviour on local machine and using alpine docker image with the latest version
2. when I gzipped file with 1M repeatable lines (resulting archive size was about 5M) everything was fine -> issue could be connected to the archive size not amount of lines

### PR DESCRIPTION
**Pull request by jszwedko**
Fixes #7061

It appears that AWS's ALB logging gzip's multi-part files which we were
only reading the first part of. I tested that `MultiGzDecoder` works on
simple gzip files so I figured we should switch to it everywhere. It was
already being used by the `file` source.

Ref: https://github.com/rust-lang/flate2-rs/issues/265#issuecomment-820525328

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

### CODE REVIEW

**Comment by binarylogic** [comment]
cc @afoninsky if you want to review / take this for a spin.

**Review by lukesteensen — APPROVED** [review]

File: src/sources/aws_s3/mod.rs [hunk 1]
```diff
@@ -167,8 +167,16 @@ fn s3_object_decoder(
     match compression {
         Auto => unreachable!(), // is mapped above
         None => Box::new(r),
-        Gzip => Box::new(bufread::GzipDecoder::new(r)),
-        Zstd => Box::new(bufread::ZstdDecoder::new(r)),
+        Gzip => Box::new({
+            let mut decoder = bufread::GzipDecoder::new(r);
+            decoder.multiple_members(true);
```
Comment (bruceg) [inline src/sources/aws_s3/mod.rs:172]: TIL That's handy!

File: src/sources/aws_s3/mod.rs [hunk 2]
```diff
@@ -290,6 +298,24 @@ mod integration_tests {
         test_event(key, Some("gzip"), None, None, buffer, logs).await;
     }
 
+    #[tokio::test]
+    async fn s3_process_message_multipart_gzip() {
```
Comment (bruceg) [inline src/sources/aws_s3/mod.rs:302]: Any thoughts on adding a multipart zstd test as well?

**Review by bruceg — APPROVED** [review]

File: src/sources/aws_s3/mod.rs (hunk 2 above)
Comment (jszwedko) [reply src/sources/aws_s3/mod.rs:302]: 👍 I added in 4dac43502

**Pull request merged**